Adds docs fixing a broken TEMPLATE pipeline of an exisiting pipeline (#574)
nf-core/nf-co.re: markdown/developers/sync.md

be created to merge updates in to your pipeline for you.

That's it, you're done! **Congratulations!**

# Fixing a broken TEMPLATE branch

> Sometimes mistakes can occur during resolving of conflicts during a template merge, typically via the GitHub interface, where the changes during the conflict
> resolution is committed to `TEMPLATE` rather than the `dev` branch itself.

Reviewer: ```suggestion
If merge conflicts are resolved via the GitHub interface instead of after pulling changes to a fork as described above, the commit history from the `dev` branch will be merged into `TEMPLATE`.
```

This leads to complex problems in later `TEMPLATE` merges as the later updated `TEMPLATE` branch removes all the pipeline-specific files that were accidentality included in problematic merge, resulting in many (in some cases >100!) of files to resolve conflicts in.

> If during one of the automated syncs you see you have an usually high number of changed files you can check whether `dev` was accidentally merged into `TEMPLATE` by looking at your repository's GitHub Network Graph under the 'Insights' tab. You can look through the pipeline history to see if there exists a commit described as 'Merge branch `dev` into `TEMPLATE`'.

Reviewer: ```suggestion
If during one of the automated syncs you see you have an usually high number of changed files you can check whether `dev` was accidentally merged into `TEMPLATE` by looking at how many commits the `TEMPLATE` branch has (should be in the range of 5-15ish, depending on how many template updates your pipeline has had). You can also look at your repository's GitHub Network Graph under the _"Insights"_ tab or even look through the `TEMPLATE` branch commit history to see if there is a commit described as 'Merge branch `dev` into `TEMPLATE`'.
```

> If so, the easiest solution is to start your TEMPLATE branch from scratch.

Reviewer 2: ```suggestion
If so, the easiest solution is to start your `TEMPLATE` branch from scratch.
```

To do this, clone your repository to your local machine, ensuring you have pulled both `TEMPLATE` and `dev` branches.

> Next, retrieve the commit hash when the original cookie-cutter template was used to generate pipeline i.e. with `nf-core create`.

Reviewer: ```suggestion
Next, retrieve the commit hash when the original nf-core template was used to generate pipeline i.e. with `nf-core create`.
```

> Next, retrieve the commit hash when the original cookie-cutter template was used to generate pipeline i.e. with `nf-core create`.

Reviewer: Instead of loads of line breaks, can this be formatted in a bullet pointed list please? I don't know about anyone else, but I find this easier to read and it takes up less space 😅

Assuming you originally started with the nf-core template, you can simply look at your git log from within your repository:

```bash
git log --reverse
```

The first commit will then typically represent the original template.

> Next we can delete the `TEMPLATE` branch with:

Reviewer: Instead of deleting and recreating, it might be easier to just reset?

```bash
git checkout TEMPLATE
git reset --hard <hash of first commit after nf-core create>
git push --force
```

```bash
git branch -D TEMPLATE
```

Now we can checkout the original pipeline setup commit:

```bash
git checkout <hash of first commit after nf-core create>
```

From this commit, we can generate a 'fresh' `TEMPLATE` branch as originally made (and without the accidently merged contents of `dev`):

```bash
git branch TEMPLATE
```

With this generated we can switch back to `dev`, and run `nf-core sync` to do it's magic and get the latest version of the template.

:warning: do not follow the `nf-core` instructions once `sync` is completed!

```bash
git checkout dev
nf-core sync .
```

Now, the `TEMPLATE` branch should in a clean state on your local machine. To replace this on the remote, you can do a force push:

```bash

> git push origin TEMPLATE --force

Reviewer: Need to consider the fact that most people will be working on their personal fork. Need to also force-push to the upstream remote.

```bash
git remote add upstream <git address for main nf-core repo>
git push upstream TEMPLATE --force
```

Reviewer: Worth adding lots of warnings that people are overwriting git history here in a way that is irreversible. So need to be very careful to do the correct branches etc!

```

This will then replace the broken `TEMPLATE` branch on GitHub with a nice clean one, which can be viewable by checking the commit history.

> With this, you're now ready to re-make the pull request from `TEMPLATE` into `dev`, and locally manually resolve conflicts (if required) following the git instructions [above](#merge-template-into-main-branches) or that GitHub provides at the bottom of the PR.

Reviewer: Don't tell people to use the GitHub merge conflicts tool - that's what got them into this mess in the first place!! 🤣 

```suggestion
With this, you're now ready to re-make the pull request from `TEMPLATE` into `dev`, and locally manually resolve conflicts (if required) following the git instructions [above](#merge-template-into-main-branches).
```

be created to merge updates in to your pipeline for you.

That's it, you're done! **Congratulations!**

# Fixing a broken TEMPLATE branch

> Sometimes mistakes can occur during resolving of conflicts during a template merge, typically via the GitHub interfacte, where the changes during the conflict

Reviewer 2: ```suggestion
Sometimes mistakes can occur during resolving of conflicts during a template merge, typically via the GitHub interface, where the changes during the conflict
```

resolution is committed to `TEMPLATE` rather than the `dev` branch itself.

> This leads to complex problems in later `TEMPLATE` merges as the later updated TEMPLATE branch removes all the pipeline-specific files that were accidently included in problematic merge, resulting in many (in some cases >100!) of files to resolve conflicts in.

Reviewer 2: ```suggestion
This leads to complex problems in later `TEMPLATE` merges as the later updated `TEMPLATE` branch removes all the pipeline-specific files that were accidentality included in problematic merge, resulting in many (in some cases >100!) of files to resolve conflicts in.
```

> If during one of the automated syncs you see you have an unsually high number of changed files you can check whether `dev` was accidently merged into `TEMPLATE` by looking at your repository's GitHub Network Graph under the 'Insights' tab. You can look through the pipeline history to see if there exists a commit described as 'Merge branch 'dev' into TEMPLATE.

Reviewer 2: ```suggestion
If during one of the automated syncs you see you have an unsually high number of changed files you can check whether `dev` was accidently merged into `TEMPLATE` by looking at your repository's GitHub Network Graph under the 'Insights' tab. You can look through the pipeline history to see if there exists a commit described as Merge branch 'dev' into `TEMPLATE`.
```
